### PR TITLE
rose bush: improve tasks and jobs filtering

### LINKED ISSUE #918: rose bush: other issues
Continue from #857.

Resolved or to be resolved:
- [x] Modify install documentation to mention new dependencies.
- [x] Task re-run not handled correctly (only latest run shown).
- [x] Task name globs not working?
- [x] Display Rose version.

Moved to new issue:
- [x] Job status labels should be consistent with those used in Cylc,
    e.g. `succeeded`, `failed`.
- [x] Adjust the styling so times are aligned.
- [x] Sort by runtime, status.
- [x] More use of top bar for search boxes/options e.g. switch user and/or suite.
- [x] Pretty print Cylc and Rose configuration files in view mode.
- [x] Print file path in view mode.
- [x] Plot suite graph.
- [x] Integrate ability to browse other users' suite outputs into rosie go
- [x] JSON tests
- [x] Download cycle job log tar files.
- [x] Group display by cycles or tasks.
- [x] User preference.
- [x] Suite summary mode - display summary of task job status.
- [x] Single start point with Rosie web service?

### LINKED ISSUE #1873: rose bush: display *jobs* filtered by state
At present rose bush can only filter by final task state so it can be hard to locate _jobs_ of interest which have, e.g. failed.

https://github.com/metomi/rose/pull/1155 looked to do this but was closed down at the time for various pertinent reasons. With the recent cylc database changes we should now be able to make this change which has been on the TODO list for a while now (although we seem to have lost the associated issue).

### PR DESCRIPTION
**Pull request by matthewrmshin**
Cycles list now displays number of tasks and jobs.

Cycles list now provides download links to those cycles with a `log/job-CYCLE.tar.gz`.

Jobs list can now be filtered by job status combinations and individual
cylc task statuses.

Job list `cycles` filter can now use syntax such as `<CYCLE` (currently `before CYCLE`), `>CYCLE` (currently `after CYCLE`).

Improve display for screens with smaller width.

Rose Bush related logic from `rose.suite_engine_procs.cylc` is moved to
`rose.bush_dao`.
- 1st step towards decoupling Rose Bush from the rest of Rose. (cylc/cylc#1052)
- Remove rose job logs DB logic.

Close #918. Close #1873.

### CODE REVIEW

**Comment by matthewrmshin** [comment]
@arjclark @oliver-sanders please review.

**Comment by arjclark** [comment]
Some initial feedback on interface:

 * as we now have the ability to filter by job status, I'd make the "active" link from the cycles page point to active jobs rather than tasks (you don't really care about the other jobs for a task at that point)
 * can you update the "active", "success" and "fail" tooltips now read task/job on the end accordingly to bring in line with "failed jobs" linking to a page for failed jobs
 * not clear what the moved last active column means in the new interface (was clear in the old and I understand its doing the same thing). Either an addition to the tooltip, similar style to old interface or column header should be added - up to you which.
 * I'd rather we had check boxes for the select by job status. If nothing else, this ensures interface consistency with what is done immediately above for task status

**Comment by matthewrmshin** [comment]
> I'd rather we had check boxes for the select by job status. If nothing else, this ensures interface consistency with what is done immediately above for task status.

I agree (sort of), but the queries for the different combos of job statuses will be very difficult to write, because the job status is not stored as a single value like the task status.

At least the interface is consistent with the ordering drop down.

**Comment by matthewrmshin** [comment]
> The count is also wrong (pre-existing problem though I think) as it is counting the number of jobs associated with an active task rather than the number of active tasks/jobs

I can't see why it would do this, as it queries only the `task_states` table for the number of active tasks.

**Comment by arjclark** [comment]
>>    The count is also wrong (pre-existing problem though I think) as it is counting the number of jobs associated with an active task rather than the number of active tasks/jobs

> I can't see why it would do this, as it queries only the task_states table for the number of active tasks.

No idea either - the only thing I can think of is that the suite I'm looking at is cycling through so quickly that this is just the way the numbers are coming up between viewing the cycle counts and actually getting the running tasks page loaded...

**Comment by matthewrmshin** [comment]
I think the numbers are correct. I can prove this by temporarily adding a column for *active jobs*. A suite with only 2 retrying tasks, for example, will have no active jobs but 2 active tasks.

**Comment by matthewrmshin** [comment]
In the latest changeset, I have modified the listings to use HTML table instead of Bootstrap grid system, as suggested by @oliver-sanders yesterday. Not sure if it is better or not. Easy to revert if not.

**Comment by matthewrmshin** [comment]
(I have done enough fiddling with this branch. Please continue review when time permits.)

**Comment by arjclark** [comment]
@matthewrmshin - I think it'd be worth running this change past someone in opsuite team to make sure they're happy as its the main tool they use for monitoring suite progress.

**Comment by arjclark** [comment]
@oliver-sanders to sanity check. I'll press the button when we believe all interested parties are happy...

File: t/rose-bush/00-basic.t [hunk 1]
```diff
@@ -39,7 +39,8 @@ SUITE_NAME="$(basename "${SUITE_DIR}")"
 cp -pr "${TEST_SOURCE_DIR}/${TEST_KEY_BASE}/"* "${SUITE_DIR}"
 export CYLC_CONF_PATH=
 cylc register "${SUITE_NAME}" "${SUITE_DIR}"
-cylc run --debug "${SUITE_NAME}"
+cylc run --debug "${SUITE_NAME}" 2>'/dev/null' \
+    || cat "${SUITE_DIR}/log/suite/err" >&2
```
Comment (oliver-sanders) [inline t/rose-bush/00-basic.t]: Debug?

**Comment by matthewrmshin** [comment]
Branch rebased.

File: lib/html/template/rose-bush/taskjobs.html [hunk 2]
```diff
@@ -22,65 +22,84 @@
...
-              />
-            </div>
-            <div class="col-md-6">
-              <input id="tasks" type="text" name="tasks" class="form-control"
-              title="Task names globs"
-              placeholder="Task names globs"
-{% if tasks -%}
-              value="{{tasks|escape}}"
+{% for key, name, value in [
+    ("cycles", "Cycles: &gt;CYCLE|&lt;CYCLE|CYCLE-GLOB ...", cycles),
```
Comment (oliver-sanders) [inline lib/html/template/rose-bush/taskjobs.html]: `>CYCLE|<CYCLE|CYCLE-GLOB ...` is somewhat cryptic.

**Comment by oliver-sanders** [comment]
For some reason the 'Task Jobs List' item in the navbar fails to highlight when you are on the taskjobs page.

File: lib/html/static/css/rose-bush.css [hunk 3]
```diff
@@ -105,3 +105,12 @@ pre {
      -moz-border-radius: 6px 0 6px 6px;
           border-radius: 6px 0 6px 6px;
 }
+
+@-moz-document url-prefix() {
+  fieldset { display: table-cell; }
```
Comment (oliver-sanders) [inline lib/html/static/css/rose-bush.css]: Do other browsers not need this rule also?

File: lib/html/template/rose-bush/taskjobs.html (hunk 2 above)
Comment (matthewrmshin) [inline lib/html/template/rose-bush/taskjobs.html]: String should now be improved.

**Comment by matthewrmshin** [comment]
> For some reason the 'Task Jobs List' item in the navbar fails to highlight when you are on the taskjobs page.

Should now be fixed.

**Comment by arjclark** [comment]
All good here. Everyone seems happy :grin:

**Pull request merged**